lib/vagrant-orchestrate/action/setcredentials.rb

```diff
@@ -0,0 +1,88 @@
+module VagrantPlugins
+  module Orchestrate
+    module Action
+      class SetCredentials
+        def initialize(app, _env)
+          @app = app
+        end
+
+        # rubocop:disable Metrics/AbcSize
+        def call(env)
+          @machine = env[:machine]
+
+          if @machine.config.orchestrate.credentials
+            @ui = env[:ui]
+            config_creds = @machine.config.orchestrate.credentials
+            (username, password) = retrieve_creds(config_creds)
+
+            # Apply the credentials to the machine info, or back out if we were unable to procure them.
+            if username && password
+              apply_creds(username, password)
+            else
+              @ui.warn <<-WARNING
+Vagrant-orchestrate could not gather credentials for machine #{@machine.name}. \
+Continuing with default credentials."
+              WARNING
+            end
+          end
+
+          @app.call env
+        end
+
+        def retrieve_creds(config_creds)
+          # Use environment variable overrides, or else what was provided in the config file
+          config_creds.username = ENV["VAGRANT_ORCHESTRATE_USERNAME"] || config_creds.username
+          config_creds.password = ENV["VAGRANT_ORCHESTRATE_PASSWORD"] || config_creds.password
+
+          # Use credentials file to any username or password that is still undiscovered
+          check_creds_file(config_creds) unless config_creds.username && config_creds.password
+
+          # Only prompt if allowed by config
+          if config_creds.prompt
+            config_creds.username ||= prompt_username
+            config_creds.password ||= prompt_password
+          end
+
+          [config_creds.username, config_creds.password]
+        end
+
+        def apply_creds(username, password)
+          configs = [@machine.config.winrm, @machine.config.ssh]
+          configs.each do |config|
+            next unless config
+            config.username = username
+            config.password = password
+          end
+        end
+
+        def prompt_username
```

Comment:
For Windows boxes (or Linux boxes in Likewise), the username will need the domain prepended. I've been doing it in my own Vagrantfiles with:

```ruby
  best_guess = ENV['USER']
  best_guess = "#{ENV['USERDOMAIN']}\\#{ENV['USERNAME']}" if ENV['USERDOMAIN']
  ask("Please enter your username:"){ |q| q.default = best_guess }
```